_transcripts/050-rust-1.56-1.57.md

```diff
@@ -0,0 +1,1302 @@
+actually. So old editions still get all the new stuff that they can, which is
+pretty much everything. There's only a few things that old editions can't
+generally get. And so in this case around 52, sometime earlier this year, I
+think it was that version, somewhere around there. Every edition of Rust got the
+ability to use `into_iter` on arrays. In most contexts. There was only one
+context where you couldn't, and it was a pretty important context, which is in
+the `for` loops.
+
+And now the new edition has, uniformly everywhere across the board, the ability
+to use `IntoIterator` on arrays. And so, old editions are kind of— they're
+featured, their full featured mostly, without— if a bit inconsistent, whereas
+the new edition is totally consistent in its ability to use this `IntoIterator`
+on arrays. And so it's kind of an example of how you don't need to lock people
+on old editions out of getting new things, you can still give them new things,
+even if nice isn't— you know, the most nice or consistent form of those things.
+
+__Jon__: The next one, I thought, is worth a little bit more discussion. So this
+is around or-patterns and `macro_rules`. So if you write a `macro_rules` defined
+macro, so a declarative macro in your code, and one of the arguments you try to
+take is using the fragment specifier type `:pat` for pattern. It used to be in
+past editions that this would not match or-patterns, so if you have like
+`A | B`. Whereas in this new edition, they will match `A | B`. And the reason
+for this was that in the old editions you could only use `A | B` at the top
+level of a pattern. So if you write `match foo`, in that first, like, the top
+level of the arm, you could say `Some(1) | Some(2)`, but you couldn't in the
+past write `Some(1 | 2)`, where the `1` or `2` is inside of the `Some`. And
+therefore it wasn't right for the macro pattern type to include ors, because it
+would mean that— the user might give you a pattern that contains an or, but the
+macro might put it in a place where it sort of ends up being a sub-pattern where
+or isn't allowed. But now, because Rust, since I think 1.53, now supports or
+inside of nested patterns as well, it was realized in the 2021 edition, we
+should just make it so that the `:pat` fragment specifier type in macros just
+includes ors as well. So that all macros that that mention these patterns allow
+users to use the full power of pattern matching. And then there's sort of a
+specific version of it called, I think "pat name", is that what they used? Let
+me double-check— `:pat_param`, it's called. That matches only the part without
+ors. And so this is one example where, if you move your crate from 2018 to 2021
+you want to make sure that any macro rules you have that allow patterns as
+arguments, that you are actually okay with the patterns including ors. Because
+in the 2021 edition they will now do so by default. Which is almost always what
+you want.
```

Comment:
This could use a quick look, to make sure the formatting is right. (audio 11:13)

Reply:
Looks right to me!